internal/controller/accessrequest_controller.go

```diff
@@ -156,6 +162,52 @@ func (r *AccessRequestReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 	return result, nil
 }
 
+type AccessRequestConflictError struct {
+	message string
+}
+
+func (e *AccessRequestConflictError) Error() string {
+	return e.message
+}
+
+func NewAccessRequestConflictError(msg string) *AccessRequestConflictError {
+	return &AccessRequestConflictError{
+		message: msg,
+	}
+}
+
+// Validate will verify if there are existing AccessRequests for the same
+// user/app/role already in progress.
+func (r *AccessRequestReconciler) Validate(ctx context.Context, ar *api.AccessRequest) error {
+	arList, err := r.findAccessRequestsByUserAndApp(ctx,
+		ar.GetNamespace(),
+		ar.Spec.Subject.Username,
+		ar.Spec.Application.Name,
+		ar.Spec.Application.Namespace)
+	if err != nil {
+		return fmt.Errorf("error finding AccessRequests by user and app: %w", err)
+	}
+	for _, arResp := range arList.Items {
+		// skip if it is the same AccessRequest
+		if arResp.GetName() == ar.GetName() &&
+			arResp.GetNamespace() == ar.GetNamespace() {
+			continue
+		}
+		// skip if the request is for different role template
+		if arResp.Spec.RoleTemplateName != ar.Spec.RoleTemplateName {
+			continue
+		}
+		// if the existing request is pending, granted, or empty (not reconciled yet)
+		// then the new request is a duplicate and must be rejected
+		if arResp.Status.RequestState == api.GrantedStatus ||
```

Comment:
Hm... what do we do if RequestState is `invalid`? Do we need to short-circut further up to avoid even considering that resource?

Reply:
Added a validation at the begging of the reconciliation loop to stop processing if the reconciliation was already concluded.